### ISSUE TITLE
abort does not abort

### ISSUE THREAD

**Issue by odcinek** [chef/ohai#516]
`abort` from inside plugin seems broken, message is being printed, but the run continues.
```
chef (12.0.3)
ohai (8.0.1)
```

**Comment by jaymzh** [comment]
Some context here - we need to able to - from an Ohai plugin, cause Ohai to completely fail and thus the chef run to fail. This used to be - in Ohai 6 - accomplished with `abort` (as opposed to `fail` which is just caught by the Ohai framework). This breaks us pretty badly. CC: @mcquin

**Comment by sean-horn** [comment]
This could be worked around in a recipe added anywhere in the run_list like `recipe[check-ohai-fail-run]` below

It checks for `node['awesome']['fatalerror']`

    raise "ouch, it hurts" if node['awesome']['fatalerror']

This flag would have been added to the node data by whatever caused a desired failure state in the Ohai plugin, like this in an Ohai7 or 8 plugin, for example

```
  collect_data(:default) do
    # Assign the value 100 to the :level key of awesome
    awesome[:level] =  100
    # Assign the value "Sriracha" to the :sauce key of awesome
    awesome[:sauce] = "Sriracha"

    if true
      awesome[:fatalerror] = true
    end

  end
```

With the following result
```
sudo chef-client -r recipe[dummy1],recipe[dummy2],recipe[dummy3],recipe[dummy],recipe[check-ohai-fail-run]
Starting Chef Client, version 12.2.1
resolving cookbooks for run list: ["dummy1", "dummy2", "dummy3", "dummy", "check-ohai-fail-run"]
Synchronizing Cookbooks:
  - dummy1
  - dummy2
  - dummy
  - check-ohai-fail-run
  - dummy3
Compiling Cookbooks...

================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/check-ohai-fail-run/recipes/default.rb
================================================================================

RuntimeError
------------
ouch, it hurts

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/check-ohai-fail-run/recipes/default.rb:33:in `from_file'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/check-ohai-fail-run/recipes/default.rb:

 26:  template_r = run_context.resource_collection.find(:template => '/tmp/template.file')
 27:  template_r.checksum = nil
 28:  end
 29:  action :nothing
 30:  notifies :create, 'template[/tmp/template.file]', :immediately
 31:  end
 32:
 33>> raise "ouch, it hurts" if node['awesome']['fatalerror']
 34:


Running handlers:
[2015-04-01T03:06:00+00:00] ERROR: Running exception handlers
Running handlers complete
[2015-04-01T03:06:00+00:00] ERROR: Exception handlers complete
[2015-04-01T03:06:00+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
Chef Client failed. 0 resources updated in 7.140054895 seconds
[2015-04-01T03:06:01+00:00] ERROR: ouch, it hurts
[2015-04-01T03:06:01+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by jaymzh** [comment]
No, it can't. Because its triggered in an attribute file... and attributes files are not loaded in runlist order, so while we have technically have a check in this in an early attribute file, someone called a function in another attribute file that accesses this data, which broke Chef on the hosts where this ohai data was missing, but without the nice error message, and instead in an odd way.

**Comment by sean-horn** [comment]
Ahh, ok.

This should do it then in the plugin's `collect_data` method.
I know it's hideous, but it does stop the run. `exit` will not stop the run in this version of Ohai, I believe because it is producing an exception in the same family as the one `#abort` produces

    exit!(123)

**Comment by jaymzh** [comment]
Wait. We're not calling `exit`, we're calling `abort` because in Ohai 6, that was the explicitly documented way to kill the Chef run from within Ohai.

I have no problem switching our code to `exit!` with a non-zero value... but is that a random hack, or is that the new "blessed" way?

**Comment by danielsdeleo** [comment]
IMO it's a bug and we should just make it work like before. FWIW, calling `exit!` will always exit in ruby since it exits the process immediately and skips all exception handling and such. `abort` is more or less an alias for `exit` which raises a `SystemExit` exception, which you can catch but usually shouldn't.

**Comment by danielsdeleo** [comment]
I would guess the problem is here: https://github.com/chef/ohai/blob/8aea6701b5680b146a3f5a94eb34f9caf5c861b6/lib/ohai/runner.rb#L51 we probably need a `rescue SystemExit; raise` clause right before that.

**Comment by astlock** [comment]
The same situation in
```
Chef: 11.16.4
Ohai: 7.4.0
```
Whether there is a temporary workaround here?

**Comment by jaymzh** [comment]
`exit!` is the work around we are currently using. It's a hack, but it works.

**Comment by MathildeBarr** [comment]
cc/ @andy-dufour @irvingpop

**Comment by chefsalim** [comment]
Fixed with PR#625
